### PR TITLE
[AIRFLOW-5831] Prod image support. Depends on [AIRFLOW-5704] [AIRFLOW-5842] [AIRFLOW-5828]

### PR DESCRIPTION
**Pull request by potiuk**
NOTE FOR REVIEWERS: It's still a Draft so it has no full description/functionality yet. If you are looking at this PR - please take a look only at the last commit as it is based on several not yet merged PRs. #5704 #6499 #6500 

Unlike CI image, the production image is optimised for size
rather than build time.

It is usable as base image for any Docker-based deployments
of Airflow (including Kubernetes image)

This production image is built automatically every time master
is merged. This way we make sure that production image is
always in mergeable state. It will also automatically use the
latest released patch-level of python images.

Make sure you have checked _all_ steps below.

### Jira

- [ ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-XXX
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [ ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

**Comment by ryw** [comment]
72 changed files hard to see the work on prod docker - is there some issue w the PR?

**Comment by potiuk** [comment]
It is based on some previous (not yet merged PRs) so  the prod image is just the last commit I have not updated the descrription yet as it is a Draft PR still. Just take a look at the last commit.

**Comment by potiuk** [comment]
The current "Prod image support" is https://github.com/apache/airflow/pull/6501/commits/700582e8b189652125181a147a2d1372e43e3714 (but it will change after rebases).

File: Dockerfile [hunk 1]
```diff
@@ -141,7 +151,8 @@ RUN mkdir -pv /usr/share/man/man1 \
       vim \
     && apt-get autoremove -yqq --purge \
     && apt-get clean \
-    && rm -rf /var/lib/apt/lists/*
+    && rm -rf /var/lib/apt/lists/* \
+    ;
 
 ENV HADOOP_DISTRO="cdh" HADOOP_MAJOR="5" HADOOP_DISTRO_VERSION="5.11.0" HADOOP_VERSION="2.6.0" \
```
Comment (svivier-orange) [inline Dockerfile:157]: Why those aren't passed as ARG?

File: Dockerfile [hunk 2]
```diff
@@ -232,10 +243,6 @@ RUN mkdir -pv ${AIRFLOW_HOME} \
     mkdir -pv ${AIRFLOW_HOME}/dags \
     mkdir -pv ${AIRFLOW_HOME}/logs
 
-# Increase the value here to force reinstalling Apache Airflow pip dependencies
-ARG PIP_DEPENDENCIES_EPOCH_NUMBER="1"
-ENV PIP_DEPENDENCIES_EPOCH_NUMBER=${PIP_DEPENDENCIES_EPOCH_NUMBER}
-
 # Optimizing installation of Cassandra driver
 # Speeds up building the image - cassandra driver without CYTHON saves around 10 minutes
 ARG CASS_DRIVER_NO_CYTHON="1"
```
Comment (svivier-orange) [inline Dockerfile:248]: Duplicate with line 114

File: Dockerfile [hunk 3]
```diff
@@ -355,21 +348,135 @@ COPY .coveragerc .rat-excludes .flake8 pylintrc LICENSE MANIFEST.in NOTICE CHANG
...
+ENV HOME=/home/airflow
+
+# Set writing bytecode back in prod environment
+ENV PYTHONDONTWRITEBYTECODE=""
+
+WORKDIR ${AIRFLOW_HOME}
+
+EXPOSE 8080
+
+ENTRYPOINT ["/home/airflow/.local/bin/dumb-init", "--", "/entrypoint.sh"]
```
Comment (svivier-orange) [inline Dockerfile]: This entrypoint is notcompatible with the current helm chart because a lot of bash cmd is used, and therefore it will fail .

**Comment by potiuk** [comment]
Noooo

File: .travis.yml [hunk 4]
```diff
@@ -66,6 +66,10 @@ jobs:
       python: "3.6"
       stage: test
       script: travis_wait 30 "./scripts/ci/ci_run_airflow_testing.sh" tests.integration.kubernetes
+    - name: "Build PROD image"
+      python: "3.5"
+      stage: test
+      script: ./scripts/ci/local_ci_build_prod_image.sh
```
Comment (mik-laj) [inline .travis.yml:72]: Where can I find this script? I can't find it on your branch.

**Comment by potiuk** [comment]
Please don't

**Comment by potiuk** [comment]
Closed in favour of #7832

**Pull request closed without merging**